### PR TITLE
Performant import of production data

### PR DESCRIPTION
**Pull request by Tirokk**
> [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-05-01T10:36:27Z" title="Wednesday, May 1st 2019, 12:36:27 pm +02:00">May 1, 2019</time>_
_Merged <time datetime="2019-05-07T21:08:54Z" title="Tuesday, May 7th 2019, 11:08:54 pm +02:00">May 7, 2019</time>_
---

@appinteractive thanks for pointing out `split`. You just saved me some
days of work to refactor the import statements to use CSV instead of
JSON files.

@Tirokk when I enter `:schema` in Neo4J web UI, I see the following:
```
:schema
Indexes
   ON :Badge(id) ONLINE
   ON :Category(id) ONLINE
   ON :Comment(id) ONLINE
   ON :Post(id) ONLINE
   ON :Tag(id) ONLINE
   ON :User(id) ONLINE

No constraints
```

So I temporarily removed the unique constraints on `slug` and added
plain indices on `id` for all relevant node types. We cannot omit the
`:Label` unfortunately, neo4j does not allow this. So I had to add all
indices for all known node labels instead.

With indices the import finishes in:
```
Time elapsed: 351 seconds
```
:tada:

@appinteractive when I keep the unique indices on slug, I get an error
during import that a node with label `:User` and slug `tobias` already
exists. Ie. we have unqiue constraint violations in our production data.

@mattwr18 @ulfgebhardt @ogerly I started the application on my machine
on the production data and it turns out that the index page
http://localhost:3000/ takes way to long. Visiting my profile page at
http://localhost:3000/profile/5b1693daf850c11207fa6109/robert-schafer
is fine, though. Even pagination works. When I visit a post page with
not too many comments, the application is fast enough, too:
http://localhost:3000/post/5bbf49ebc428ea001c7ca89c/neues-video-format-human-connection-tech-news

### Issues

- [X] None

### Checklist

- [X] None

### How2Test

- [X] None

### Todo

- [X] None

### CODE REVIEW

**Comment by Tirokk** [comment]
> [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-05-01T11:25:17Z" title="Wednesday, May 1st 2019, 1:25:17 pm +02:00">May 1, 2019</time>_

---


Did some query profiling on the generated cypher statement of [this graphQL query](https://github.com/Human-Connection/Human-Connection/blob/master/webapp/pages/index.vue#L94):


plan


I cannot really read and understand the plan above :point_up: but I removed parts of the graphQL query and it seems that those counts and nested queries are very expensive. Especially everything in `author { ... }`.

**Comment by Tirokk** [comment]
> [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-05-01T11:43:37Z" title="Wednesday, May 1st 2019, 1:43:37 pm +02:00">May 1, 2019</time>_

---


Here are our duplicates in the production data:

## Users (34 duplicates)
```cypher
match (u:User)
match (u2:User)
where u.slug = u2.slug and u <> u2
return u.slug, u2.slug
```

|           |           | 
|-----------|-----------| 
| u.slug    | u2.slug   | 
| andre     | andre     | 
| pat       | pat       | 
| jurgen    | jurgen    | 
| manuel    | manuel    | 
| christoph | christoph | 
| maria     | maria     | 
| marion    | marion    | 
| sebastian | sebastian | 
| tobias    | tobias    | 
| susanne   | susanne   | 
| marvin    | marvin    | 
| wolf      | wolf      | 
| lars      | lars      | 
| andy      | andy      | 
| sonja     | sonja     | 
| guido     | guido     | 
| max       | max       | 
| tobias    | tobias    | 
| sebastian | sebastian | 
| andy      | andy      | 
| marvin    | marvin    | 
| manuel    | manuel    | 
| jurgen    | jurgen    | 
| max       | max       | 
| lars      | lars      | 
| sonja     | sonja     | 
| andre     | andre     | 
| maria     | maria     | 
| christoph | christoph | 
| wolf      | wolf      | 
| guido     | guido     | 
| susanne   | susanne   | 
| pat       | pat       | 
| marion    | marion    | 


The positive message: There seem to be no duplicates for label types `Comment`, `Post` or `Category`.

@appinteractive I had a quick look into the returned records. The duplicates have distinct email addresses, so I believe the user accounts are not used by the same person. The users are neither `disabled` nor `deleted`. 

Do you have an idea why we have duplicate users in the database?

Do you think it is safe to "repair" those unique constraint violations manually?

**Comment by Tirokk** [comment]
> [appinteractive](https://github.com/appinteractive) **Authored by [appinteractive](https://github.com/appinteractive)**
_<time datetime="2019-05-01T13:01:37Z" title="Wednesday, May 1st 2019, 3:01:37 pm +02:00">May 1, 2019</time>_

---


@roschaefer glad that I could help with the import! 

The performance issue on the newsfeed comes from the sorting by date! At least that’s the issue I encountered but did not had a solution. Maybe we need to use another date format there or some index!?

The dublicated slugs are strange, maybe they are related to invites!? Maybe appand a number to the slug of the least active users?

File: deployment/legacy-migration/maintenance-worker/migration/mongo/import.sh [hunk 1]
```diff
@@ -9,16 +9,17 @@ echo "MONGODB_DATABASE         ${MONGODB_DATABASE}"
...
+mkdir -p /tmp/mongo-export/
 
 ssh -4 -M -S my-ctrl-socket -fnNT -L 27018:localhost:27017 -l ${SSH_USERNAME} ${SSH_HOST}
 
 for collection in "categories" "badges" "users" "contributions" "comments" "follows" "shouts"
 do
-  mongoexport --host localhost -d ${MONGODB_DATABASE} --port 27018 --username ${MONGODB_USERNAME} --password ${MONGODB_PASSWORD} --authenticationDatabase ${MONGODB_AUTH_DB} --db ${MONGODB_DATABASE} --collection $collection --out "/tmp/mongo-export/$collection.json"
+  mongoexport --db ${MONGODB_DATABASE} --host localhost -d ${MONGODB_DATABASE} --port 27018 --username ${MONGODB_USERNAME} --password ${MONGODB_PASSWORD} --authenticationDatabase ${MONGODB_AUTH_DB} --collection $collection  --collection $collection --out "/tmp/mongo-export/$collection.json"
+  mkdir -p /tmp/mongo-export/splits/$collection/
+  split -l 1000 -a 3 /tmp/mongo-export/$collection.json /tmp/mongo-export/splits/$collection/
```
Comment (Tirokk) [inline deployment/legacy-migration/maintenance-worker/migration/mongo/import.sh:22]: > [appinteractive](https://github.com/appinteractive) **Authored by [appinteractive](https://github.com/appinteractive)**
_<time datetime="2019-05-01T13:05:23Z" title="Wednesday, May 1st 2019, 3:05:23 pm +02:00">May 1, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -9,16 +9,17 @@ echo "MONGODB_DATABASE         ${MONGODB_DATABASE}"
 echo "MONGODB_AUTH_DB          ${MONGODB_AUTH_DB}"
 echo "-------------------------------------------------"
 
-[ -z "$SSH_PRIVATE_KEY" ] || create_private_ssh_key_from_env
 
 rm -rf /tmp/mongo-export/*
-mkdir -p /tmp/mongo-export
+mkdir -p /tmp/mongo-export/
 
 ssh -4 -M -S my-ctrl-socket -fnNT -L 27018:localhost:27017 -l ${SSH_USERNAME} ${SSH_HOST}
 
 for collection in "categories" "badges" "users" "contributions" "comments" "follows" "shouts"
 do
-  mongoexport --host localhost -d ${MONGODB_DATABASE} --port 27018 --username ${MONGODB_USERNAME} --password ${MONGODB_PASSWORD} --authenticationDatabase ${MONGODB_AUTH_DB} --db ${MONGODB_DATABASE} --collection $collection --out "/tmp/mongo-export/$collection.json"
+  mongoexport --db ${MONGODB_DATABASE} --host localhost -d ${MONGODB_DATABASE} --port 27018 --username ${MONGODB_USERNAME} --password ${MONGODB_PASSWORD} --authenticationDatabase ${MONGODB_AUTH_DB} --collection $collection  --collection $collection --out "/tmp/mongo-export/$collection.json"
+  mkdir -p /tmp/mongo-export/splits/$collection/
+  split -l 500 /tmp/mongo-export/$collection.json /tmp/mongo-export/splits/$collection/
```


Slit is limited to max 676 parts as it used two letters, you should append -a 3 which used three letters. Also go with maybe 1000 lines?

Comment (Tirokk) [reply deployment/legacy-migration/maintenance-worker/migration/mongo/import.sh:22]: > [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-05-07T19:49:15Z" title="Tuesday, May 7th 2019, 9:49:15 pm +02:00">May 7, 2019</time>_

---


Done!

**Review by Mogge — APPROVED** [review]
Looks great @roschaefer!! glad it's working out
great tip on the `split` @appinteractive

**Pull request closed without merging**